### PR TITLE
java implementation of vscode debug adapter data types

### PR DESCRIPTION
**Pull request by testforstephen**

### CODE REVIEW

**Review by andxu — APPROVED** [review]

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Events.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+package org.eclipse.jdt.ls.debug.adapter;
+
+import java.util.HashMap;
+
+/**
+ * The event types defined by VSCode Debug Protocol.
+ */
+public class Events {
+    public static class DebugEvent {
+        public String type;
```
Comment (yaohaizh) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Events.java:21]: public field?

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Events.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+        public DebugEvent(String type) {
+            this.type = type;
+        }
+    }
+
+    public static class DebugProtocolCallbacks {
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Events.java]: Remove unused one.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Results.java [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+/**
+ * The response content types defined by VSCode Debug Protocol.
+ */
+public class Results {
+    /**
+     * This monomorphic class is used to return results from a debugger request
+     * or to return errors. In addition events can be attached that are fired
+     * after the request results have been returned to the caller.
+     */
+    public static class DebugResult {
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Results.java]: Seems this is not necessary. DispatcherReponse/DispatcherEvent can be used directly.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Types.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        }
+    }
+
+    public static class Capabilities {
+        public boolean supportsConfigurationDoneRequest;
+        public boolean supportsFunctionBreakpoints;
+        public boolean supportsConditionalBreakpoints;
+        public boolean supportsHitConditionalBreakpoints;
+        public boolean supportsEvaluateForHovers;
+        public boolean supportsStepBack;
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Types.java]: We won't support this..

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Types.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        public boolean supportsSetVariable;
+        public boolean supportsDelayedStackTraceLoading;
+        public List<ExceptionBreakpointFilter> exceptionBreakpointFilters;
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Types.java]: Not supported for now. Remote this line.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Types.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+    public static class Capabilities {
+        public boolean supportsConfigurationDoneRequest;
+        public boolean supportsFunctionBreakpoints;
+        public boolean supportsConditionalBreakpoints;
+        public boolean supportsHitConditionalBreakpoints;
+        public boolean supportsEvaluateForHovers;
+        public boolean supportsStepBack;
+        public boolean supportsSetVariable;
+        public boolean supportsDelayedStackTraceLoading;
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Types.java]: Not supported.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Types.java [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+                return;
+            }
+            this.filter = value;
+        }
+    }
+
+    public static class Capabilities {
+        public boolean supportsConfigurationDoneRequest;
+        public boolean supportsFunctionBreakpoints;
+        public boolean supportsConditionalBreakpoints;
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Types.java]: Not supported.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Types.java [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+        public boolean supportsHitConditionalBreakpoints;
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Types.java]: Not supported.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Types.java [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+            if (!Filter_All.equals(value) && !Filter_UserUnhandled.equals(value)) {
+                return;
+            }
+            this.filter = value;
+        }
+    }
+
+    public static class Capabilities {
+        public boolean supportsConfigurationDoneRequest;
+        public boolean supportsFunctionBreakpoints;
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Types.java]: Not supported.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Types.java [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+         */
+        public void setFilter(String value) {
+            if (!Filter_All.equals(value) && !Filter_UserUnhandled.equals(value)) {
+                return;
+            }
+            this.filter = value;
+        }
+    }
+
+    public static class Capabilities {
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Types.java:217]: Add supportTerminateDebuggee.

**Review by akaroml — CHANGES_REQUESTED** [review]

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Events.java (hunk 1 above)
Comment (testforstephen) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/Events.java:21]: yes, the field should be accessed by gson serialization and de-serialization.

**Review by akaroml — APPROVED** [review]
Perfect. Thank you!
Please **Squash and merge**

**Pull request merged**